### ISSUE TITLE
lexical search vs. semantic search

### ISSUE THREAD

**Issue by art-in** [art-in/nano_search#6]
compare quality of lexical search (tf-idf and bm25) and semantic search (embedding vectors + ANN)

for lexical search use existing nano_search and tantivy engines  
for semantic search add new engine  

- [ ] pick evaluation dataset (documents, queries, doc-query relevances)
- [ ] add semantic search engine
  - [ ] pick embedding model
  - [ ] pick vector database
- [ ] evaluate and compare search quality

**Comment by art-in** [comment]
pick embedding runtime and model

using [ONNX runtime](https://github.com/microsoft/onnxruntime) through [ort](https://github.com/pykeio/ort) and  [fastembed](https://github.com/Anush008/fastembed-rs) crates

- ONNX is one of most popular model formats
  - ONNX runtime allows model inference without original training engine, like pytorch or tensorflow 
  - onnx-dir often present in huggingface's model files, but if not - pytorch/tensorflow models can be converted to onnx
- `ort` crate is a rust wrapper around original c++ ONNX runtime
- `fastembed` crate is a wrapper around `ort`, which adds downloading and caching of target models

---

currently experimenting with different models

started with `sentence-transformers/all-MiniLM-L6-v2` - small sentence embedding model.  
search precision on small datasets (CISI, NFCorpus) is little worse than with lexical search/BM25
 
need to experiment more with other larger datasets and models

**Comment by art-in** [comment]
results of search quality evaluation with averaged nDCG@10 metric on [BEIR](https://github.com/beir-cellar/beir?tab=readme-ov-file#beers-available-datasets) datasets:

|       -                          | nano         | tantivy        | vector        |
| ---                             | ---            | ---              | ---              |
| nfcorpus                   | 0.295       | 0.300          | **0.311**   |
| quora                        | 0.723        | 0.737          | **0.875**  |
| scifact                      | **0.657**  | 0.648         | 0.654         |
| scidocs                     | 0.150        | 0.146          | **0.216**   |
| arguana                    | 0.356        | 0.338         | **0.359**   |
| trec-covid                | 0.526        | **0.559**   | 0.444         |
| fiqa                           | 0.222         | 0.232         | **0.361**   |
| webis-touche2020 | 0.362         | **0.368**         | 0.161          |

<img width="1422" height="423" alt="Image" src="https://github.com/user-attachments/assets/9b61e42a-ecf1-4686-82db-cbccfef28ad3" />

- nano and tantivy use lexical search with BM25 ranking  
- vector search uses [sentence-transformers/all-MiniLM-L6-v2](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2) embedding model and L2/Euclidean similarity

**Comment by art-in** [comment]
key findings:

- embeddings are not that small

  depending on vector dimentionality and storing method, embeddings can be larger than source documents  

  e.g. embeddings of CISI documents are larger than documents themselves:  
  - avg doc size is ~1KB
  - embedding size is = 384 dimentions * float32 = ~1.5KB

- embeddings may be very slow

  inferencing embedding model on CPU takes so much time, it's not even practical without GPU acceleration

  99% of vector indexing time with CPU-only approach is calculating embeddings

  e.g. indexing 8.8M MSMARCO docs on macbook pro m2 with 10 CPU cores without GPU acceleration takes
    - ~3 minutes with nano search (lexical indexing)
    - ~1 minute with tantivy search (lexical indexing)
    - ~17 hours with vector search

- CoreML may slow down model inference

  ONNX runtime offers [CoreML](https://developer.apple.com/documentation/coreml) executor, which can choose GPU or Neural Engine depending on model operations and available hardware for accelerating model inference on apple devices

  but on macbook pro m2 it either makes inference of MiniLM models 7x times slower than CPU-only, or doesn't activate at all due to ["CoreML does not support input dim > 16384" warnings](https://github.com/microsoft/onnxruntime/issues/19543)

  because of this I've skipped evaluating big datasets with > 1M docs

- [nDCG](https://en.wikipedia.org/wiki/Discounted_cumulative_gain) metric is more reliable than precision/recall

  there's no ideal metric of search quality in IR, [all of them](https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)) show results from different perspectives. but unlike precision/recall, nDCG also takes doc positions in search results into account

  e.g. vector search shows lower precision compared to lexical search, but higher nDCG on nfcorpus dataset. it means vector search usually puts fewer relevant docs into result, but same time scores them higher. so it has better chance to retrieve relevant doc from, say, top-1, compared to lexical search. and nDCG helps to highlights that

- lexical search (inverted index + BM25) is still looks good

  most of the time vector search gives higher quality results, than lexical search.  
  but (1) not much higher, and (2) it takes much more hardware resources for both indexing and searching

  also, [BEIR](https://openreview.net/pdf?id=wCu6T5xFjeJ) study shows that to really squeeze quality out of vector search it requires more sofisticated approaches, than simple embeddings+ANN. e.g. fine-tune models to catch domain specifics, late-interaction, re-ranking (BM25 + rerank based on vectors similarity), etc. and even with that, best method still results with only +11% higher avg quality

  overall, taking all the factors into account, like (1) amount and price of hardware resources, (2) speed of indexing and search, (3) index disk size, (4) quality of search results -- lexical search still looks competitive
